### PR TITLE
Use detach c api instead of stop_gradient

### PR DESCRIPTION
**Pull request by stu1130**
## Description ##
detach & stop_gradient are doing the same thing (correct me if I am wrong). The former one is for dynamic graph while the later is for static graph

### CODE REVIEW

**Comment by stu1130** [comment]
also should we use another set of name for this kind of gradient manipulation? attachGradient & detachGradient seems confuses the user with NDManager attach & detach

**Comment by stu1130** [comment]
the other way is to change NDManager attach & detach. Since most people are coming from Python side, they will be used to using detach and they will see it’s available to use but it will do a completely different thing by Marcos

**Comment by zachgk** [comment]
detach and stopGradient should be different things. If attachGradient says that you should track gradients for an NDArray, detachGradient says you shouldn't track it. So, they mutate the NDArray.

The stopGradient is to make a copy of the NDArray. But, copies usually will still propagate gradients back to the original while stopGradient does not. It can then be used for magnifying gradients or other gradient manipulations when used instead of or in conjunction with the original gradient propagating array.

**Comment by stu1130** [comment]
In PyTorch, requires_grad=False have the limitation on leaf variables where you have to use detach(). https://discuss.pytorch.org/t/understanding-of-requires-grad-false/39765 so It looks like they are the same?

**Comment by zachgk** [comment]
Yeah, so the relationship seems to be:

| DJL | PyTorch | MXNet |
| ---- | -----| ---- |
| attachGradient | require_grad(True)| attach_grad('write')|
| detachGradient | require_grad(False)|attach_grad(null)|
| stopGradient| detach |stopGrad and maybe also ndarray.detach|

**Comment by stu1130** [comment]
> Yeah, so the relationship seems to be:
> 
> DJL	PyTorch	MXNet
> attachGradient	require_grad(True)	attach_grad('write')
> detachGradient	require_grad(False)	attach_grad(null)
> stopGradient	detach	stopGrad and maybe also ndarray.detach

If that is the case, I propose we should remove the case 2 and keep only attachGradient & detachGradient. Does that sound good?

**Comment by zachgk** [comment]
Case 2 is the same as freezing parameters so I think we still want it. I like requiresGrad(Boolean) for case 1 and 2. It expresses the user goal (this NDArray requires the gradients so that it can be optimized).

I am not as much a fan of detach because it "detaches" from the computation graph. But, that is a weird comment and it is hard to understand what the implications of it are. Stop gradient which stops gradients from propagating is a bit more straightforward. It is also closer to the similar operation of scaleGradient which magnifies particular gradients more than back-propagation normally would.

**Comment by stu1130** [comment]
> Case 2 is the same as freezing parameters so I think we still want it. I like requiresGrad(Boolean) for case 1 and 2. It expresses the user goal (this NDArray requires the gradients so that it can be optimized).
> 
> I am not as much a fan of detach because it "detaches" from the computation graph. But, that is a weird comment and it is hard to understand what the implications of it are. Stop gradient which stops gradients from propagating is a bit more straightforward. It is also closer to the similar operation of scaleGradient which magnifies particular gradients more than back-propagation normally would.

DJL Parameter already have an attribute called requiresGrad to cover freezing params so we can still get rid of case 2. But in general, I am also fine with your proposal. @frankfliu @lanking520 @goswamig @aksrajvanshi what do you guys think?

**Comment by stu1130** [comment]
I have updated the PR to add requiresGrad(boolean)

**Review by zachgk — APPROVED** [review]
